[WGSL] Add AST node for WorkgroupSizeAttribute

tadeuzagallo opened the pull request:
#### 1b0340486392c92f8f0899604c2e3b726be22c00
<pre>
[WGSL] Add AST node for WorkgroupSizeAttribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=251362">https://bugs.webkit.org/show_bug.cgi?id=251362</a>
&lt;rdar://problem/104820892&gt;

Reviewed by Myles C. Maxfield.

Add support for `@workgroup_size(\d)` annotations

* Source/WebGPU/WGSL/AST/AST.h:
* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/AST/ASTWorkgroupSizeAttribute.h: Copied from Source/WebGPU/WGSL/AST/AST.h.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseAttribute):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/259613@main">https://commits.webkit.org/259613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/853705a0df1bdbc5b9d57e1b0cff16901d9265b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38166 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114531 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5274 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97589 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114458 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111026 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12016 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95001 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39500 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93882 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26635 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81162 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7686 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27995 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7781 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4575 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13835 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47539 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6631 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9569 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/AST/ASTWorkgroupSizeAttribute.h:
```diff
@@ -0,0 +1,55 @@
+
+namespace WGSL::AST {
+
+class WorkgroupSizeAttribute final : public Attribute {
```
2 questions:
1. Why is this its own node rather than simply being an enum field in a `class Attribute`?
2. Attributes are not recursive, so why make them AST nodes in the first place? Why not describe them via an elaborate `Variant` typedef instead of having them be heap-allocated and traversed by the Visitor?

Context: I made Attributes be AST nodes in WHLSL and I ended up regretting it - it cost more in implementation lines of code than it gained. None of the passes actually ended up using it without _also_ using the context of its owner. And if you need the context of the owner in order to do anything useful with an attribute, might as well just have the owner's visitor function just inspect any present attributes directly, rather than asking the Visitor to recurse to them.

tadeuzagallo replied:
I'm mostly onboard, not sure about variants, but I agree they don't necessarily need to be traversed. I think I want to refactor a lot of this, but I want to sync with @djg before any drastic changes, so for now I'm just sticking to the current style and adding just enough to it to unblock myself.

tadeuzagallo committed:
[WGSL] Add AST node for WorkgroupSizeAttribute
https://bugs.webkit.org/show_bug.cgi?id=251362
<rdar://problem/104820892>

Reviewed by Myles C. Maxfield.

Add support for `@workgroup_size(\d)` annotations

* Source/WebGPU/WGSL/AST/AST.h:
* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/AST/ASTWorkgroupSizeAttribute.h: Copied from Source/WebGPU/WGSL/AST/AST.h.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseAttribute):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/259613@main
